### DISCUSSION TITLE
get_toc(simple=False) return 'to' point coordinate is not based on top-left origin

### DISCUSSION THREAD

**Discussion by charosen** [pymupdf/PyMuPDF#3413] [Q&A]
### Description of the bug

i have a pdf, with outlines(titles) and content below:

```
1.1 Hello World

1.1.1. first step to hello world

content
```

and i want to extract all the outline(titles) and their coordinates in page.

when i use `get_toc(simple=False)`, fitz return a toc list:

```
[[1,
  '1.1 Hello world',
  1,
  {'kind': 4,
   'xref': 41631,
   'page': 0,
   'to': Point(0.0, 761.8583),
   'zoom': 0.0,
   'nameddest': '_OPENTOPIC_TOC_PROCESSING_d13321e25969',
   'collapse': True,
   'color': (0.0, 0.0, 0.0)}],
 [2,
  '1.1.1 first step to hello world',
  1,
  {'kind': 4,
   'xref': 41632,
   'page': 0,
   'to': Point(0.0, 731.8583),
   'zoom': 0.0,
   'nameddest': '_OPENTOPIC_TOC_PROCESSING_d13321e25972',
   'collapse': True,
   'color': (0.0, 0.0, 0.0)}],
...
]
```

the returned **'to' points** is not based on **top-left origin**, but **bottom-left origin**, because `1.1 Hello world` is above `1.1.1 first step to hello world'`, but **Point(0.0, 761.8583)** is greater than **Point(0.0, 731.8583),**

it seems like pdf coordinates, **not (py)mupdf coordinates.**

how to covert those toc **'to' points** to **top-bottom coordinates**.

### How to reproduce the bug

import fitz

document = fitz.open('mypdf.pdf')

toc = document.get_toc(simple=False)

toc results:
```
[[1,
  '1.1 Hello world',
  1,
  {'kind': 4,
   'xref': 41631,
   'page': 0,
   **'to': Point(0.0, 761.8583),**
   'zoom': 0.0,
   'nameddest': '_OPENTOPIC_TOC_PROCESSING_d13321e25969',
   'collapse': True,
   'color': (0.0, 0.0, 0.0)}],
 [2,
  '1.1.1 first step to hello world',
  1,
  {'kind': 4,
   'xref': 41632,
   'page': 0,
   **'to': Point(0.0, 731.8583),**
   'zoom': 0.0,
   'nameddest': '_OPENTOPIC_TOC_PROCESSING_d13321e25972',
   'collapse': True,
   'color': (0.0, 0.0, 0.0)}],
...
]
```

### PyMuPDF version

1.24.1

### Operating system

Linux

### Python version

3.9

**Comment by JorjMcKie** [comment]
You did not provide the reproducing file.

**Comment by charosen** [comment]
> You did not provide the reproducing file.

sorry, i could not upload mypdf file for some reason. 

However, it is pretty clear that **'to' point** in toc is based on `bottom-left` origin, not `top-left` origin.

i simply want to convert **'to' points** to `top-left` coordinates.

**Comment by JorjMcKie** [comment]
It is not all clear:
What are we even looking at? Where do the "**" come from?
The TOC entries seem to point to named destinations - are there errors in the PDF? Or in our code?
Did the PDF creator **_want_** to point to the bottom left point 🤷‍♂️?
Have you tried to look at the PDF's names dictionary?

Again: without the file in question we are already wasting time.

**Comment by JorjMcKie** [comment]
Maybe you simply had a question and just wanted to know how to do coordinate transformation?
In that case you shouldn't have submitted an error report but a post in Discussions.

**Comment by charosen** [comment]
> It is not all clear: What are we even looking at? Where do the "**" come from? The TOC entries seem to point to named destinations - are there errors in the PDF? Or in our code? Did the PDF creator **_want_** to point to the bottom left point 🤷‍♂️? Have you tried to look at the PDF's names dictionary?
> 
> Again: without the file in question we are already wasting time.

Sorry for the "**" signs,  i just want to get bolded fonts, and i already delete them.

my question is:

`get_toc(simple=False)` returns a Point(0.0, 761.8583) for `1.1 Hello World`, and a Point(0.0, 731.8583) for `1.1.1. first step to hello world`.

`1.1 Hello World` is above `1.1.1. first step to hello world`, however, Point(0.0, 761.8583) is greater than Point(0.0, 731.8583), which is not based on pymupdf top-left coordinates.

**Comment by JorjMcKie** [comment]
Ok - to make **some** progress, I transferring this thread to Discussions, and we can continue there.

**Reply by charosen** [reply]
ok, sorry for my mis-submitting to bugs.

**Reply by charosen** [reply]
I want to know:
1. how to convert `get_toc` "to" Points back into top-left coordinates, in other words, convert pdf coordinates back to (py)mupdf coordinates?
2. any attribute in `get_toc results` could indicate whether the "to" point is `top-left` coordinates or `bottom-left` coordinates? when i should do convertion ?

**Comment by JorjMcKie** [comment]
The two TOC entries obviously point to **named** destinations. You can extract (but not set / update) a PDF's defined symbolic names as a Python [dictionary](https://pymupdf.readthedocs.io/en/latest/document.html#Document.resolve_names) via `names = doc.resolve_names()`. As documented, the "to" values there are in PDF coordinates - (0, 0) is bottom-left. This is also the "to" value in the TOC link.
You need the page object to transform this point value to MuPDF coordinates like this: `fitz.Point(0.0, 731.8583) * page.transformation_matrix`.

**Reply by charosen** [reply]
ok, i got it.

for my first question:  `fitz.Point(0.0, 731.8583) * page.transformation_matrix` will do the conversion

**Reply by JorjMcKie** [reply]
This does a matrix multiplication point * matrix, resulting of course in a point.

**Reply by charosen** [reply]
for my second question:

based on documentation:
![image](https://github.com/pymupdf/PyMuPDF/assets/12933334/7dabdbab-2edf-4bb8-b57b-332dd9bfbe2d)




when the dest is `named destinations`, in other words, kind == 4, i should do the conversion ? 
when the dest is  `Points to a place in this document`, kind == 1, i should not do it ?

**Reply by JorjMcKie** [reply]
correct

**Reply by charosen** [reply]
thanks for patiently guiding!!! 
i try to fix my codes now!! 

thanks again !!!

**Reply by JorjMcKie** [reply]
you are welcome
